r/R/arrow-tabular.R

```diff
@@ -211,6 +211,25 @@ head.ArrowTabular <- head.ArrowDatum
 #' @export
 tail.ArrowTabular <- tail.ArrowDatum
 
+#' @export
+na.fail.ArrowTabular <- function(object, ...){
+  na_count <- sum(purrr::map_int(object$columns, ~.x$null_count))
+  if(na_count > 0){
+    stop("missing values in object")
+  }
+  object
+}
+
+#' @export
+na.omit.ArrowTabular <- function(object, ...){
+  not_na <- purrr::map(object$columns, ~!is.na(.x))
+  not_na_agg <- purrr::reduce(not_na, `&`)
+  object$Filter(not_na_agg)
```

Comment:
A few thoughts:

1. It would be nice to build up a single array_expression rather than eagerly evaluate each part. It doesn't change anything today, but soon it will allow some optimizations in the Arrow C++ library, and it would help if we're already set up for that.
2. Building the expression also lets you use "is_valid", which is more efficient than `negate(is_null(x))`
3. This seems like the kind of thing that should ultimately be implemented in the C++ library--it's useful and there are optimizations to be made by dealing with this at a lower level. Can you please make a JIRA?
4. We import `purrr::map` so you don't need to `::` it.

```suggestion
  not_na <- map(object$columns, ~build_array_expression("is_valid", .x))
  not_na_agg <- Reduce("&", not_na)
  object$Filter(eval_array_expression(not_na_agg))
```